### PR TITLE
Reduce boilerplate in applySteps:

### PR DESCRIPTION
**Pull request by seelabs**
## High Level Overview of Change

When a new transactor is added, there are several places in applySteps that need to be modified. This patch refactors the code so only one function needs to be modified.

### Type of Change

- [ x] Refactor (non-breaking change that only restructures code)

### CODE REVIEW

File: src/ripple/app/tx/impl/applySteps.cpp
Comment (ckeshava) [inline src/ripple/app/tx/impl/applySteps.cpp:144]: I had not come across the usage of the `template` keyword, I learnt a new construct :) thanks.

File: src/ripple/app/tx/impl/applySteps.cpp
Comment (ckeshava) [inline src/ripple/app/tx/impl/applySteps.cpp:211]: Why is this `assert` statement necessary? The exception will be caught and throw without the assert too.

File: src/ripple/app/tx/impl/applySteps.cpp
Comment (ckeshava) [inline src/ripple/app/tx/impl/applySteps.cpp:320]: Does the developer need to specify `invoke_apply.template operator<Some_Txn_Type>()` the type explicitly like this? 

So you have effectively collected the all the common "functionality" from `invoke_apply`, `invoke_preflight_helper`, `invoke_preclaim` and `invoke_calculateBaseFee` and placed them under `with_txn_type`. Am I understanding this correctly?

File: src/ripple/app/tx/impl/applySteps.cpp
Comment (nbougalis) [inline src/ripple/app/tx/impl/applySteps.cpp:67]: Why not `F&&`?

**Review by nbougalis — COMMENTED** [review]
Nice! I like this a lot. It's certainly advanced C++ and some people may go "wha... ?!?" when they read it, but it does make it less error-prone to add new transaction types.

Good job!

File: src/ripple/app/tx/impl/applySteps.cpp
Comment (nbougalis) [reply src/ripple/app/tx/impl/applySteps.cpp:211]: The intention is to have servers running in debug mode to "break" into the debugger if they encounter an unknown transaction. The wisdom of this is... debatable, but this bit was already there before.

Comment (ckeshava) [reply src/ripple/app/tx/impl/applySteps.cpp:211]: ok, thanks for the clarification

File: src/ripple/app/tx/impl/applySteps.cpp
Comment (seelabs) [reply src/ripple/app/tx/impl/applySteps.cpp]: There's no reason it has to be `const`. Fixed in `e253734ede [fold] Use universal ref for callback`, and yeah, I tend to prefer universal refs for callbacks myself. I don't know why I wrote it as a `const&` (although it doesn't really cause harm either).

File: src/ripple/app/tx/impl/applySteps.cpp
Comment (seelabs) [reply src/ripple/app/tx/impl/applySteps.cpp:330]: I may be misunderstanding your question, but `invoke_apply` is called like this: `invoke_apply(ctx);`. It's a regular function and doesn't have an `operator()`. 

`with_txn_type` factors out the `switch` statement that used to live in those four places, and handles the variability between those functions by using a callback. So now when a new transaction is added we need to change one place rather than four.

File: src/ripple/app/tx/impl/applySteps.cpp
Comment (ckeshava) [inline src/ripple/app/tx/impl/applySteps.cpp:275]: @seelabs  How does this lambda deduce the type T? T needs to be explicitly specified by the caller of `invoke_apply` or any of the other three functions right?

File: src/ripple/app/tx/impl/applySteps.cpp
Comment (mvadari) [inline src/ripple/app/tx/impl/applySteps.cpp:196]: Shouldn't these comments stay?

File: src/ripple/app/tx/impl/applySteps.cpp
Comment (mvadari) [reply src/ripple/app/tx/impl/applySteps.cpp:275]: C++20 adds support for templated lambdas: https://ggulgulia.medium.com/c-20-lambda-expression-and-template-syntax-7f6ae2f9e7cd
`with_txn_type` will provide the type to the lambda based on the value of `tx.getTxnType`.

Comment (mvadari) [inline src/ripple/app/tx/impl/applySteps.cpp:275]: Why do the `view` and `tx` parameters not need to be captured in the lambda?

File: src/ripple/app/tx/impl/applySteps.cpp
Comment (mvadari) [reply src/ripple/app/tx/impl/applySteps.cpp:330]: `invoke_apply` is a normal function, it's just doing complex things under the hood.

File: src/ripple/app/tx/impl/applySteps.cpp
Comment (seelabs) [reply src/ripple/app/tx/impl/applySteps.cpp:196]: I don't find the comment that helpful - I think the code is clear without it. But I also don't mind leaving it in. Fixed in `eb4547ad04 [fold] Re add comment`

File: src/ripple/app/tx/impl/applySteps.cpp
Comment (seelabs) [reply src/ripple/app/tx/impl/applySteps.cpp:275]: The `[&]` will default capture all the needed variables by ref. So `view` and `tx` _are_ both captured. This could have been written as `[&view, &tx]` but there's no reason to call attention to what variables are captured.

Comment (ckeshava) [reply src/ripple/app/tx/impl/applySteps.cpp:275]: On a related note, will the const-ness of these refs be preserved through the lambda? I remember watching a lecture about this a long time ago, but I'm not able to discern it now.

Comment (ckeshava) [reply src/ripple/app/tx/impl/applySteps.cpp:275]: thanks! this makes a lot more sense now 👍

Comment (ckeshava) [reply src/ripple/app/tx/impl/applySteps.cpp:275]: i found some solutions for the const question here: https://stackoverflow.com/questions/3772867/lambda-capture-as-const-reference

**Review by mvadari — COMMENTED** [review]
I'm not entirely sure this refactor will play nicely with plugins, but it doesn't really hurt anything, because the plugin code can just continue to do what it's doing, with separate code in each of the methods.

**Review by mvadari — APPROVED** [review]

**Comment by intelliot** [comment]
@seelabs - do you you have an opinion on whether this should go in 2.0 (with XChainBridge and lots of other changes), or wait for 2024?

I hope this is a safe refactor, but you never know ;)

**Review by ckeshava — APPROVED** [review]

**Comment by ckeshava** [comment]
@mvadari What do you mean by plug-in code? Is it a different PR?

**Comment by mvadari** [comment]
> @mvadari What do you mean by plug-in code? Is it a different PR?

It will be in a future PR - it's still WIP.

**Comment by seelabs** [comment]
@intelliot I don't think we need to delay this. I think we can merge this into the next release.

**Review by gregtatcam — APPROVED** [review]
👍 Nice update! One minor suggestion to consider - change `std::enable_if_t` in `consequences_helper()` to `requires`. It just makes it a bit more readable.

**Comment by seelabs** [comment]
Force pushed to resync with my local branch. I also disabled maintainers from pushing to this branch. It's easier for me to track things if I'm the only one who pushes.

**Comment by intelliot** [comment]
note: do not merge until after PR author applies `Passed` label

**Comment by seelabs** [comment]
@gregtatcam Nice suggestion! I agree the "requires" is nicer (it's not part of vocabulary yet so reached for `enable_if` just out of habbit). Fixed in `c887c0bfd0 [fold] Use requires clause in place of enable_if`. Can you do me a favor and give a thumbs up after you take a quick look so I know it's OK with you.

**Comment by gregtatcam** [comment]
> @gregtatcam Nice suggestion! I agree the "requires" is nicer (it's not part of vocabulary yet so reached for `enable_if` just out of habbit). Fixed in `c887c0bfd0 [fold] Use requires clause in place of enable_if`. Can you do me a favor and give a thumbs up after you take a quick look so I know it's OK with you.

Sure, will do.

**Review by gregtatcam — APPROVED** [review]
👍 LGTM

**Review by ximinez — APPROVED** [review]
I freakin' love this. Approved as-is.

...But I am going to throw you a curve ball. What do you think of getting rid of the exception type, and using an `UnknownTxn` placholder transactor instead? To me, the big advantage is that it puts all the error conditions into one place (the class) instead of spread around the `catch` clauses.

Here's the change: https://github.com/seelabs/rippled/compare/refactor-apply-steps...ximinez:rippled:sd-4710-apply-steps

**Comment by seelabs** [comment]
@ximinez (I really wish I could respond to top level comment in github). That's an interesting design! It is nice that it gets rid of the try/catch. My biggest objection is it's a non-standard way to handle errors. It's relatively easy to look at the current code and understand what's going on - exceptions mean errors, there's a try/catch block...yeah, yeah, yeah we've seen it a million times. Your design is codes the error as concrete class that returns the correct error code. It's really clever! I kinda like it! But it's also a way of handling errors in a way we don't usually do in rippled. I'm a little hesitant to introduce another error handling mechanism.

Having said all that, I do like what you did. What do you think about getting this in now and maybe do your patch as a later PR? (I'll admit another minor concern is another round of reviews, and I'm trying to get this merged; But that shouldn't be a strong criteria).

**Comment by seelabs** [comment]
Squashed and rebased onto the latest develop

**Pull request merged**
